source/containers/document-pii-detection/parsers/html_parser.py

```diff
@@ -0,0 +1,152 @@
+    ]
+
+    def _visible(self, element):
+        """Used to filter text elements that have invisible text on the page.
+        """
+        if element.name in self._disallowed_names:
+            return False
+        elif re.match(u'<!--.*-->', six.text_type(element.extract())):
+            return False
+        return True
+
+    def _inline(self, element):
+        """Used to check whether given element can be treated as inline
+        element (without new line after).
+        """
+        if element.name in self._inline_tags:
+            return True
+        return False
+
+    def _find_any_text(self, tag):
+        """Looks for any possible text within given tag.
+        """
+        text = ''
+        if tag is not None:
+            text = six.text_type(tag)
+            text = re.sub(r'(<[^>]+>)', '', text)
+            text = re.sub(r'\s', ' ', text)
+            text = text.strip()
+        return text
+
+    def _parse_tables(self, soup):
+        """Returns array containing basic informations about tables for ASCII
+        replacement (look: _replace_tables()).
+        """
+        tables = []
+        for t in soup.find_all('table'):
+            t_dict = {'width': 0, 'table': t, 'trs': [], 'col_width': {}}
+            trs = t.find_all('tr')
+            if len(trs) > 0:
+                for tr in trs:
+                    tr_dict = []
+                    tds = tr.find_all('th') + tr.find_all('td')
+                    if len(tds) > 0:
+                        for i, td in enumerate(tds):
+                            td_text = self._find_any_text(td)
+                            length = len(td_text)
+                            if i in t_dict['col_width']:
+                                t_dict['col_width'][i] = max(
+                                    length,
+                                    t_dict['col_width'][i]
+                                )
+                            else:
+                                t_dict['col_width'][i] = length
+                            tr_dict.append({
+                                'text': td_text,
+                                'colspan': int(td.get('colspan', 1)),
+                            })
+                        t_dict['trs'].append(tr_dict)
+                for col in t_dict['col_width']:
+                    t_dict['width'] += t_dict['col_width'][col]
+                tables.append(t_dict)
+        return tables
+
+    def _replace_tables(self, soup, v_separator=' | ', h_separator='-'):
+        """Replaces <table> elements with its ASCII equivalent.
+        """
+        tables = self._parse_tables(soup)
+        v_sep_len = len(v_separator)
+        v_left_sep = v_separator.lstrip()
+        for t in tables:
+            html = ''
+            trs = t['trs']
+            h_length = 1 + (v_sep_len * len(t['col_width'])) + t['width']
+            head_foot = (h_separator * h_length) + "\n"
+            html += head_foot
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears that you are concatenating a string inside a loop. Concatenating immutable sequences will always results in a new object. This means that building up a sequence by repeated concatenation will have a quadratic runtime cost in the total sequence length. For efficiency and to get a linear runtime cost, consider appending values to a list and concatenating that list using [str.join()](https://docs.python.org/3/library/stdtypes.html#str.join) or [bytes.join()](https://docs.python.org/3/library/stdtypes.html#bytes.join) method.

[Learn more](https://docs.python.org/3/faq/programming.html#what-is-the-most-efficient-way-to-concatenate-many-strings-together)

Similar issue at line number 137.

Reply:
Resolved by concatenating using str.join()